readdy/main/model/reactions/ReactionRegistry.cpp

```diff
@@ -413,21 +405,89 @@ ReactionRegistry::addDecay(const std::string &name, const std::string &type, sca
+}
+
+namespace {
+struct FindId {
+    explicit FindId(std::string name) : name(std::move(name)) {}
+
+    std::string name;
+    Reaction::reaction_id id = 0;
+    bool found = false;
+
+    template<typename T>
+    void operator()(const T &reaction) {
+        if (reaction->name() == name) {
+            found = true;
+            id = reaction->id();
+        };
     }
-    one_educt_registry_internal[t].push_back(std::move(reaction));
-    _n_order1 += 1;
-    return id;
+};
+
+struct FindName {
+    explicit FindName(Reaction::reaction_id id) : id(id) {}
+
+    Reaction::reaction_id id;
+    std::string name = "";
+    bool found = false;
+
+    template<typename T>
+    void operator()(const T &reaction) {
+        if (reaction->id() == id) {
+            found = true;
+            name = reaction->name();
+        };
+    }
+};
 }
 
-const short ReactionRegistry::addExternal(reactions::Reaction<2> *r) {
-    two_educts_registry_external[std::tie(r->educts()[0], r->educts()[1])].push_back(r);
-    _n_order2 += 1;
-    return r->id();
+bool ReactionRegistry::reactionNameExists(const std::string &name) const {
+    FindId findId(name);
+    readdy::util::collections::for_each_value_ref(one_educt_registry_internal, findId);
+    readdy::util::collections::for_each_value_ref(one_educt_registry_external, findId);
+    readdy::util::collections::for_each_value_ref(two_educts_registry_internal, findId);
+    readdy::util::collections::for_each_value_ref(two_educts_registry_external, findId);
+    return findId.found;
+}
+
+std::string ReactionRegistry::nameOf(ReactionRegistry::reaction_id id) const {
+    FindName findName(id);
+    readdy::util::collections::for_each_value_ref(one_educt_registry_internal, findName);
+    readdy::util::collections::for_each_value_ref(one_educt_registry_external, findName);
+    readdy::util::collections::for_each_value_ref(two_educts_registry_internal, findName);
+    readdy::util::collections::for_each_value_ref(two_educts_registry_external, findName);
+    if (findName.found) {
+        return findName.name;
+    } else {
+        throw std::runtime_error(fmt::format("no reaction with id {} exists", id));
+    }
+}
+
+ReactionRegistry::reaction_id ReactionRegistry::idOf(const std::string &name) const {
+    FindId findId(name);
+    readdy::util::collections::for_each_value_ref(one_educt_registry_internal, findId);
+    readdy::util::collections::for_each_value_ref(one_educt_registry_external, findId);
+    readdy::util::collections::for_each_value_ref(two_educts_registry_internal, findId);
+    readdy::util::collections::for_each_value_ref(two_educts_registry_external, findId);
+    if (findId.found) {
+        return findId.id;
+    } else {
+        throw std::runtime_error(fmt::format("no reaction with name {} exists", name));
+    }
+}
+
+const ReactionRegistry::reactions_raw_ptr_map ReactionRegistry::allReactions() const {
+    reactions_raw_ptr_map all;
```

Comment:
I would call this `idMap` (or something alike). Also i would not call this method within the simulation loop as it makes copies out of the `orderNFlat` methods which in itself make copies already. Instead, one could use `registry.order1()` and `registry.order2()` directly.